node/cn/tracers/api.go

```diff
@@ -647,7 +654,12 @@ func (api *CommonAPI) traceBlock(ctx context.Context, block *types.Block, config
 
 		pend = new(sync.WaitGroup)
 		jobs = make(chan *txTraceTask, len(txs))
+
+		header   = block.Header()
+		blockCtx = blockchain.NewEVMBlockContext(header, newChainContext(ctx, api.backend), nil)
 	)
+	api.backend.Engine().Initialize(newChainContext(ctx, api.backend), header, statedb)
```

Comment:
In L643, you can replace StateAtBlock with StateAtTransaction(... txIndex=0 ...) to eliminate the need to Initialize(). The same goes for traceChain.

Then the whole discussion about ChainContext can be voided.

Reply:
As-is
```
traceBlock(100)
  stateAtBlock(99) -> returns the state after block 99
  for tx in txs(100)
    ApplyTransaction(tx)
```

To-be
```
traceBlock(100)
  stateAtTransaction(100, txIndex=0)
    stateAtBlock(99)
    engine.Initialize()
  for tx in txs(100)
    ApplyTransaction(tx)
```

Note that with EIP-2953, `state after block 99 = state before block 100` is no longer true. Tracer APIs want the `state before block 100` so using `stateAtTransaction` seems more appropriate.

Reply:
To verify my understanding, an outline below was your suggestion?

```
traceBlock(100)
  stateAtTransaction(100, txIndex=0)
  for tx in txs(100)
    ApplyTransaction(tx)
```

which contributes to removing the explicit `Initialize()`  call. Additionally, system transactions do not make dirty the state before executing a simulation. Correct?

Reply:
PTAL https://github.com/kaiachain/kaia/pull/121/commits/20c2c1c181573ea210211fa347040019c575e347